### ISSUE TITLE
Wrong initial value from PCAP capture

### ISSUE THREAD

**Issue by Araneidae** [PandABlocks/PandABlocks-server#37]
It seems that since trigger time capture was added to [PandABlocks-FPGA](https://github.com/PandABlocks/PandABlocks-FPGA) the first capture now returns stale data from the previous capture.

This is reported in issue https://github.com/PandABlocks/PandABlocks-FPGA/issues/129, but is actually a server kernel driver issue as reported by @EmilioPeJu in https://github.com/PandABlocks/PandABlocks-FPGA/issues/129#issuecomment-1620818318.

**Comment by Araneidae** [comment]
I don't actually see what's going on here after all.  Shouldn't `IRQ_STATUS_LENGTH(status)` be zero, which means we would be capturing an empty block anyway?  And shouldn't the timestamp capture be _inside_ the `open->stream_active` test?

If the length field is non-zero then this remains a firmware bug.

**Comment by EmilioPeJu** [comment]
The problem is that a sample count of 0 also causes an block advance

**Comment by Araneidae** [comment]
Does that cause a problem?  If there's no data ... then there's no data and the block advance just generates an empty tick on the server.  The interface is specifically designed to work like this, and the `panda_stream_read` just returns `-EAGAIN` in this case.

**Comment by EmilioPeJu** [comment]
The problem is that the FPGA doesn't switch the buffer when there is only a start event, and the servers assumes the FPGA did

**Comment by Araneidae** [comment]
Oh, I think I see.  I guess this is reflected in bit 6?  That'll be why you recommend testing bits 0, 5, 6:

| Bit | Meaning |
| --- | --- |
| 0 | End of transfer, no more interrupts, must mark stream as inactive |
| 5 | Timeout on block, block advanced |
| 6 | Block complete ... does this mean block is full? |

Is bit 6 always set on completion of block?  If so, it was redundant and bit 5 is not actually useful, suspect it's never tested.

I'll figure out what needs to be done.

**Comment by EmilioPeJu** [comment]
Any of those bit set indicate a switch in buffer, before my change, that was the only reason to have an interrupt. Now that we have the start event interrupt, if this happens in isolation, we shouldn't switch buffer. The change I suggested is something like the following patch:
```diff
diff --git a/driver/panda_stream.c b/driver/panda_stream.c
index f395264..b4d3702 100644
--- a/driver/panda_stream.c
+++ b/driver/panda_stream.c
@@ -146,6 +146,7 @@ static inline int step_index(int ix)
 #define IRQ_STATUS_DMA_ACTIVE(status)   ((bool) ((status) & 0x80))
 #define IRQ_STATUS_COMPLETION(status)   ((size_t) (((status) >> 1) & 0x0F))
 #define IRQ_STATUS_START_EVENT(status)  ((bool) ((status)  & 0x100))
+#define IRQ_NEED_NEW_BUFFER(status)     ((bool) ((status) & 0x61))


 static void advance_isr_block(struct stream_open *open)
@@ -198,8 +199,11 @@ static irqreturn_t stream_isr(int irq, void *dev_id)

     smp_rmb();

-    if (IRQ_STATUS_START_EVENT(status))
+    if (IRQ_STATUS_START_EVENT(status)) {
         ktime_get_real_ts64(&open->start_ts);
+        if (!IRQ_NEED_NEW_BUFFER(status))
+            return IRQ_HANDLED;
+    }

     if (open->stream_active)
     {

```
Unfortunately, I cannot test until I'm back from vacation, but I hope the idea is clear

**Comment by Araneidae** [comment]
PR #38 is much the same, but I check for `stream_active` before processing any interrupts, I prefer to avoid early return when reasonable, and the macro name matches the existing names more closely.

**Comment by EmilioPeJu** [comment]
Fixed by Michael in #38
